### PR TITLE
ON-1002 Admin service UI enhancements

### PR DESCRIPTION
**Pull request by ChrisJWoodcock**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Domain/Paging/PaginatedList.cs [hunk 1]
```diff
@@ -3,27 +3,66 @@
...
+        public override int FirstVisiblePage => Math.Max(Math.Min(PageIndex - (PageSetSize / 3), (TotalPages + 1) - PageSetSize), 1);
+        public override int LastVisiblePage => Math.Min(Math.Max(PageIndex + (PageSetSize / 2), PageSetSize), TotalPages);
+
+        public override int FirstVisibleItem => ((PageIndex-1) * PageSize) + 1;
+        public override int LastVisibleItem => Math.Min(PageIndex * PageSize, TotalRecordCount);
+
+        public PaginatedList<T1> Convert<T1>() where T1 : class
+        {
+            return new PaginatedList<T1>(Items.ConvertAll(p => p as T1), TotalRecordCount, PageIndex, PageSize, PageSetSize);
         }
```
Comment (cofaulco) [inline src/SFA.DAS.AssessorService.Domain/Paging/PaginatedList.cs]: This code is never used

File: src/SFA.DAS.AdminService.Web/Extensions/RouteValueDictionaryExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using Microsoft.AspNetCore.Routing;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+
+namespace SFA.DAS.AdminService.Web.Extensions
+{
+    public static class RouteValueDictionaryExtensions
+    {
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web/Extensions/RouteValueDictionaryExtensions.cs]: I struggled on how to pass this to the tag helper so I read the source code for .Net Core and that is how it was done there so I copied it.

File: src/SFA.DAS.AdminService.Web/Views/Register/_OrganisationStandardTabsPartial.cshtml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        </section>
+    }
+    @if (Model.FeedbackApplications.HasStandards)
+    {
+        <section class="govuk-tabs__panel" id="feedback">
+            <div>
+                @{await Html.RenderPartialAsync("_OrganisationStandardApplicationsPartial", @Model.FeedbackApplications);}
+            </div>
+        </section>
+    }
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web/Views/Register/_OrganisationStandardTabsPartial.cshtml]: This is a view with tabs (a single ul) with optionally the tabs showing if there would be anything in them, the sections need to be outside the ul as they are the contents - so need separate if statements for the tabs and their contents.

File: src/SFA.DAS.AdminService.Web/ViewModels/Register/RegisterViewOrganisationStandardsViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+        public OrganisationStandardsApplicationsViewModel InProgressApplications { get; set; }
+
+        public OrganisationStandardsApplicationsViewModel FeedbackApplications { get; set; }
+
+        public bool HasStandards
+        {
+            get
+            {
+                return OrganisationStandards.HasStandards || InProgressApplications.HasStandards || FeedbackApplications.HasStandards;
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web/ViewModels/Register/RegisterViewOrganisationStandardsViewModel.cs]: Done

File: src/SFA.DAS.AdminService.Web/ViewModels/Register/OrganisationStandardsApplicationsViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        public string OrganisationId { get; set; }
+
+        public PaginationViewModel<ApplicationSummaryItem> PaginationViewModel { get; set; }
+
+        public bool HasStandards
+        {
+            get
+            {
+                return (PaginationViewModel?.PaginatedList?.TotalRecordCount ?? 0) > 0;
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web/ViewModels/Register/OrganisationStandardsApplicationsViewModel.cs]: Done

File: src/SFA.DAS.AdminService.Web/ViewModels/Register/OrganisationApprovedStandardsViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+        public PaginationViewModel<OrganisationStandardSummary> PaginationViewModel { get; set; }
+
+        public bool HasStandards
+        {
+            get
+            {
+                return (PaginationViewModel?.PaginatedList?.TotalRecordCount ?? 0) > 0;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web/ViewModels/Register/OrganisationApprovedStandardsViewModel.cs]: Done

File: src/SFA.DAS.AdminService.Web/Infrastructure/CheckSessionFilter.cs [hunk 7]
```diff
@@ -18,25 +20,56 @@ public CheckSessionFilter(ILogger<CheckSessionAttribute> logger, ISessionService
...
-                context.Result = new RedirectToActionResult("SignIn", "Account", null);
-                _logger.LogInformation("Session lost, redirecting to Sign In");
+                if (checkSession.CheckSession == CheckSession.Error)
+                {
+                    context.Result =
+                        new BadRequestObjectResult("Session lost");
+
+                    _logger.LogInformation($"Session lost, error result");
+                }
+                else if (checkSession.CheckSession == CheckSession.Redirect)
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web/Infrastructure/CheckSessionFilter.cs]: Done - Renamed

File: src/SFA.DAS.AdminService.Web.Tests/Controllers/Register/Given_I_request_view_organiation_page_after_paging_organsation_standards.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+using Microsoft.AspNetCore.Mvc;
+using NUnit.Framework;
+using SFA.DAS.AdminService.Web.Controllers;
+using SFA.DAS.AdminService.Web.Extensions.TagHelpers;
+using SFA.DAS.AdminService.Web.ViewModels.Register;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.AdminService.Web.Tests.Controllers.Register
+{
+    public class Given_I_request_view_organiation_page_after_paging_organsation_standards : RegisterBase
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web.Tests/Controllers/Register/Given_I_request_view_organiation_page_after_paging_organsation_standards.cs]: Corrected in 3 places

File: src/SFA.DAS.AdminService.Web.Extensions/TagHelpers/RoleTagHelper.cs [hunk 9]
```diff
@@ -28,14 +28,23 @@ public RoleTagHelper(IHttpContextAccessor contextAccessor, IAuthorizationService
...
+                output.SuppressOutput();
+            }
+        }
 
-            Dictionary<string, bool> roleValidation = roles.ToDictionary(role => role, role => _contextAccessor.HttpContext.User.IsInRole(role));
+        public bool ShowForRoles(string roles)
+        {
+            string[] roleArray = roles.Split(',');
 
+            Dictionary<string, bool> roleValidation = roleArray.ToDictionary(role => role, role => _contextAccessor.HttpContext.User.IsInRole(role));
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web.Extensions/TagHelpers/RoleTagHelper.cs]: Agreed I didn't read the original code just shuffled it around a bit...

File: src/SFA.DAS.AssessorService.Domain/Paging/PaginatedList.cs (hunk 1 above)
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Domain/Paging/PaginatedList.cs]: This is a copy of the code from Assessor which does use it - I was trying to keep it the same - as really this whole project belongs in the Assessor Repo - there are copies of the code which are 99.5% the same between the repos

File: src/SFA.DAS.AdminService.Web/Views/Application/Section.cshtml [hunk 10]
```diff
@@ -1,12 +1,23 @@
...
     Layout = "_Layout";
 }
 
-<a class="govuk-back-link" asp-action="Sequence" asp-controller="Application" asp-route-applicationId="@Model.ApplicationId" asp-route-sequenceNo="@Model.SequenceNo">Back</a>
+<a class="govuk-back-link"
+    asp-action="Sequence"
+    asp-controller="Application"
+    asp-route-applicationId="@Model.ApplicationId"
+    asp-route-sequenceNo="@Model.SequenceNo"
+    asp-route-backAction="@Model.BackAction"
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web/Views/Application/Section.cshtml]: They render a link which should have these properties as route values; so this is correct; however have updated the methods to which these are linked are routes to receive the properties in a model class

**Pull request merged**